### PR TITLE
[Clang] fix crash by avoiding invalidation of extern main declaration during strictness checks

### LINKED ISSUE #104570: #101853 breaks `extern "C" int main...` even when -Wmain is disabled
#101853 introduces a couple of problems. I guess, this boils down to marking the declaration of main() invalid in case it's declared with extern "C", but I may be wrong here.

The first problem is a clang crash: https://gcc.godbolt.org/z/dWETErhzM

The second one is a change of behavior of code that declares main() extern "C" (see my comment https://github.com/llvm/llvm-project/commit/37ec6e5f12afe4a37872bf28f280423696f39019#r145380228). This one still needs some work to create a standalone repro for, but I suspect that both issues are related to marking the declaration invalid.

The problem with both cases is that in some codebases the non-trivial amount of cleanup is necessary for this, but Clang doesn't allow to postpone the cleanup and retain the old behavior by just ignoring the diagnostic (-Wno-main). Even if the diagnostic is not issued, the issues above still manifest.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #104570

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -264,6 +264,7 @@ Bug Fixes to C++ Support
 - Properly reject defaulted copy/move assignment operators that have a non-reference explicit object parameter.
 - Clang now properly handles the order of attributes in `extern` blocks. (#GH101990).
 - Fixed an assertion failure by preventing null explicit object arguments from being deduced. (#GH102025).
+- Fixed a crash caused by marking the function declaration as invalid during strictness checks on ``extern`` main. (#GH104570).
```
Comment (alexfh) [inline clang/docs/ReleaseNotes.rst:267]: Is it actually necessary, given that the issue was introduced in the same release?

**Review by alexfh — APPROVED** [review]
Looks good with one nit.

File: clang/lib/Sema/SemaDecl.cpp [hunk 2]
```diff
@@ -12233,7 +12233,6 @@ void Sema::CheckMain(FunctionDecl *FD, const DeclSpec &DS) {
        FD->getDeclContext()->getRedeclContext()->isTranslationUnit())) {
     Diag(FD->getLocation(), diag::ext_main_invalid_linkage_specification)
         << FD->getLanguageLinkage();
-    FD->setInvalidDecl();
     return;
```
Comment (alexfh) [inline clang/lib/Sema/SemaDecl.cpp:12236]: Should we proceed with other checks after this one?

**Review by alexfh — COMMENTED** [review]
One more comment, actually.

File: clang/lib/Sema/SemaDecl.cpp (hunk 2 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @alexfh Thanks for the review. I've removed it to proceed with other checks...

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (a-tarasyuk) [reply clang/docs/ReleaseNotes.rst]: Good question. Should this be treated as a follow-up to the main PR, or listed separately as a fix? I've removed it for now, but I can revert it if needed...

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp [hunk 3]
```diff
@@ -122,6 +119,26 @@ namespace ns2 {
...
 
+#elif TEST14
+extern "C" {
+  int main(); // expected-warning {{'main' should not be 'extern "C"'}}
+}
+
+extern "C" int main(); // expected-warning {{'main' should not be 'extern "C"'}}
+
+#elif TEST15
+extern "C" __attribute__((visibility("default"))) __attribute__((weak))
```
Comment (alexfh) [inline clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:130]: nit: This can be simplified a bit:
```
extern "C" __attribute__((visibility("default"))) __attribute__((weak))
int main();

unsigned long g() {
  return reinterpret_cast<unsigned long>(&main);
}
```
https://gcc.godbolt.org/z/j5crjfebq

Comment (a-tarasyuk) [reply clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:130]: I've updated test, thanks

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (alexfh) [reply clang/docs/ReleaseNotes.rst]: Release notes should list changes from the previous release. Since both the issue and the fix are within the same release (LLVM 20), there's no need to describe the fix separately. That's my understanding at least ;)

**Comment by alexfh** [comment]
Thanks for the fix, @a-tarasyuk! Can you merge the PR or should I do this for you?

**Comment by a-tarasyuk** [comment]
@alexfh Please merge the PR as I lack the access. Thanks

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp [hunk 4]
```diff
@@ -78,12 +80,12 @@ namespace ns {
 extern "C" struct A { int main(); }; // ok
 
 namespace c {
-  extern "C" void main(); // expected-warning {{'main' should not be 'extern "C"'}}
+  extern "C" int main(); // expected-warning {{'main' should not be 'extern "C"'}}
 }
 
 extern "C" {
   namespace Z {
-    void main(); // expected-warning {{'main' should not be 'extern "C"'}}
```
Comment (cor3ntin) [inline clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:86]: Why did you change this test?

Comment (a-tarasyuk) [reply clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:86]: To avoid unnecessary diagnostics and only validate invalid linkage for the main function, since the main function will not be interrupted by invalid usage. Should I revert it?

Comment (cor3ntin) [reply clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:86]: Yes - i think it's useful to test that `void main()` is recognized as a main function, albeit an invalid one

**Review by Sirraide — APPROVED** [review]
Ah, I should have noticed that `setInvalidDecl()` makes no sense there when I reviewed that. LGTM after fixing what Corentin pointed out.

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp (hunk 4 above)
Comment (a-tarasyuk) [reply clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp]: @cor3ntin Thanks for the feedback. I've reverted to test _void main_ functions.

**Review by cor3ntin — APPROVED** [review]

**Comment by cor3ntin** [comment]
I'll merge when CI passes, thanks
